### PR TITLE
expression: fix a compatibility issue in JSON_CONTAINS

### LINKED ISSUE #9957: JSON_CONTAIN is not same with mysql.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```SQL
drop table if exists jemp;
CREATE TABLE jemp (
    `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
    `category` JSON,
    `tags` JSON,
    PRIMARY KEY (`id`)
);
INSERT INTO `jemp` (category, tags) VALUES ('{"id": 1, "name": "TiDB"}', '[1, 2, 3]');
INSERT INTO `jemp` (category, tags) VALUES (JSON_OBJECT("id", 2, "name", "pingcap.com"), JSON_ARRAY(1, 3, 5));
SELECT * FROM jemp WHERE JSON_CONTAINS(category, 1, '$.id');
```

2. What did you expect to see?

## In MySQL
```SQL
mysql> drop table if exists jemp;
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TABLE jemp (
    ->     `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
    ->     `category` JSON,
    ->     `tags` JSON,
    ->     PRIMARY KEY (`id`)
    -> );
Query OK, 0 rows affected (0.03 sec)

mysql> INSERT INTO `jemp` (category, tags) VALUES ('{"id": 1, "name": "TiDB"}', '[1, 2, 3]');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO `jemp` (category, tags) VALUES (JSON_OBJECT("id", 2, "name", "pingcap.com"), JSON_ARRAY(1, 3, 5));
Query OK, 1 row affected (0.00 sec)

mysql> SELECT * FROM jemp WHERE JSON_CONTAINS(category, 1, '$.id');
ERROR 3146 (22032): Invalid data type for JSON data in argument 2 to function json_contains; a JSON string or JSON type is required.

mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 5.7.25-0ubuntu0.18.04.2 |
+-------------------------+
1 row in set (0.00 sec)

```

3. What did you see instead?
## In TiDB

```SQL
mysql> drop table if exists jemp;
Query OK, 0 rows affected (0.03 sec)

mysql> CREATE TABLE jemp (
    ->     `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
    ->     `category` JSON,
    ->     `tags` JSON,
    ->     PRIMARY KEY (`id`)
    -> );
Query OK, 0 rows affected (0.03 sec)

mysql> INSERT INTO `jemp` (category, tags) VALUES ('{"id": 1, "name": "TiDB"}', '[1, 2, 3]');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO `jemp` (category, tags) VALUES (JSON_OBJECT("id", 2, "name", "pingcap.com"), JSON_ARRAY(1, 3, 5));
Query OK, 1 row affected (0.00 sec)

mysql> SELECT * FROM jemp WHERE JSON_CONTAINS(category, 1, '$.id');
+----+---------------------------+-----------+
| id | category                  | tags      |
+----+---------------------------+-----------+
|  1 | {"id": 1, "name": "TiDB"} | [1, 2, 3] |
+----+---------------------------+-----------+
1 row in set (0.01 sec)

```
4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v3.0.0-beta-295-g97f88f70d
Git Commit Hash: 97f88f70d56e69d6aef584b2bd4c9ada2ce26598
Git Branch: master
UTC Build Time: 2019-03-29 07:34:36
GoVersion: go version go1.12.1 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
This PR aims to close #9957

### What is changed and how it works?
Added more args check

### Check List 

Tests 

 - Unit test
 
Code changes

 - Has exported function/method change
 
Side effects

 Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by bb7133** [comment]
PTAL @jackysp @erjiaqing

File: expression/builtin_json.go [hunk 1]
```diff
@@ -651,6 +651,19 @@ func (c *jsonContainsFunctionClass) getFunction(ctx sessionctx.Context, args []E
 	if err := c.verifyArgs(args); err != nil {
 		return nil, err
 	}
+
+	// This should actually do in `c.verifyArgs` but there's no way to inherit a method in Go.
+	evalType := args[0].GetType().EvalType()
+	if evalType != types.ETJson && evalType != types.ETString {
+		return nil, json.ErrInvalidJSONData.GenWithStack(
+			"Invalid data type for JSON data in argument 1 to function json_contains; a JSON string or JSON type is required.")
```
Comment (jackysp) [inline expression/builtin_json.go]: Should we upgrade the error type?
I found that MySQL had a special error type for it.
https://github.com/mysql/mysql-server/blob/8.0/share/errmsg-utf8.txt#L7559

Comment (bb7133) [reply expression/builtin_json.go]: Sounds reasonable, I just checked the message from MySQL 5.7

File: expression/builtin_json_test.go [hunk 2]
```diff
@@ -379,60 +379,73 @@ func (s *testEvaluatorSuite) TestJSONContains(c *C) {
...
+		if t.errMsg == "" {
 			c.Assert(err, IsNil)
 			if t.expected == nil {
 				c.Assert(d.IsNull(), IsTrue)
 			} else {
 				c.Assert(d.GetInt64(), Equals, int64(t.expected.(int)))
 			}
 		} else {
-			c.Assert(err, NotNil)
+			c.Assert(err.Error(), Equals, t.errMsg)
```
Comment (erjiaqing) [inline expression/builtin_json_test.go]: Maybe we should check the type of error?

File: expression/builtin_json_test.go [hunk 3]
```diff
@@ -379,60 +379,73 @@ func (s *testEvaluatorSuite) TestJSONContains(c *C) {
...
 	}
+	// For issue 9957: test 'argument 1 and 2 as valid json object'
+	_, err := fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(1, ``)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 1 to function json_contains; a JSON string or JSON type is required.")
+	_, err = fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(0.05, ``)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 1 to function json_contains; a JSON string or JSON type is required.")
+	_, err = fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(``, 1)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 2 to function json_contains; a JSON string or JSON type is required.")
+	_, err = fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(``, 0.05)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 2 to function json_contains; a JSON string or JSON type is required.")
```
Comment (erjiaqing) [inline expression/builtin_json_test.go]: Can we move these tests to tbl?

File: expression/builtin_json_test.go (hunk 2 above)
Comment (bb7133) [reply expression/builtin_json_test.go]: What 'type' do you mean? I didn't fully understand

File: expression/builtin_json_test.go (hunk 3 above)
Comment (bb7133) [reply expression/builtin_json_test.go]: @erjiaqing Those errors are returned by `getFunctions`, IMHO trying to integrate into `tbl` will make the codes complicated and confusing.

File: expression/builtin_json_test.go (hunk 2 above)
Comment (erjiaqing) [reply expression/builtin_json_test.go]: Maybe use `err.Equal()` is better?

File: expression/builtin_json_test.go (hunk 3 above)
Comment (erjiaqing) [reply expression/builtin_json_test.go]: Can we check them in line 427?

Comment (erjiaqing) [reply expression/builtin_json_test.go]: Or, at least, use a loop.

Comment (bb7133) [reply expression/builtin_json_test.go]: > Can we check them in line 427?
That means `tbl` need 2 error fields, which is ugly IMHO

Comment (bb7133) [reply expression/builtin_json_test.go]: > Or, at least, use a loop.

addressed, thanks!

**Comment by bb7133** [comment]
PTAL @jackysp @erjiaqing

File: expression/builtin_json.go [hunk 4]
```diff
@@ -651,6 +651,17 @@ func (c *jsonContainsFunctionClass) getFunction(ctx sessionctx.Context, args []E
 	if err := c.verifyArgs(args); err != nil {
 		return nil, err
 	}
+
+	// TODO: This should actually do in `c.verifyArgs` but there's no way to inherit a method in Go.
```
Comment (erjiaqing) [inline expression/builtin_json.go]: how about
```go
func (c *jsonContainsFunctionClass) verifyArgs(args []Expression) error {
	if err := c.baseFunctionClass.verifyArgs(args); err != nil {
        return err
	}
	if evalType := args[0].GetType().EvalType(); evalType != types.ETJson && evalType != types.ETString {
		return json.ErrInvalidJSONData.GenWithStackByArgs(1, "json_contains")
	}
	if evalType := args[1].GetType().EvalType(); evalType != types.ETJson && evalType != types.ETString {
		return json.ErrInvalidJSONData.GenWithStackByArgs(2, "json_contains")
	}
	return nil
}
```

Comment (bb7133) [reply expression/builtin_json.go]: great! thanks

**Pull request merged**